<h1 style="color:#1a237e; background-color:#ffeb3b; padding:10px; border-radius:5px; text-align:center;text-shadow: 2px 2px 5px rgba(0,0,0,0.5);">
   Brazilian E-Commerce: Data Analysis & Insights
</h1>
>

<div style="background-color:#1e1e2f; color:white; padding:15px; border-radius:10px; font-size:16px;">
    
<h2 style="color:#ffcc00;">🚀 Overview</h2>

<p>
Brazil is one of the fastest-growing e-commerce markets in the world. With a rapidly expanding internet user base, a young digital-savvy population, and increasing smartphone penetration, online shopping has become an essential part of consumer behavior in the country.
</p>

<p>
This analysis explores key trends, sales performance, customer behavior, and product category insights within the Brazilian e-commerce market. By leveraging data-driven techniques, we aim to uncover patterns that can help businesses optimize their strategies and improve customer experiences.
</p>

</div>


<h1 style="color:lightGreen; font-size:32px; font-weight:bold; background-color:#1C1C1C;padding:10px; border-radius:8px;">1.Importing Required Libraries</h1>

##### <span style="color:orange">Importing required libraries means loading external Python libraries into your script or notebook so you can use their functions and tools.

In [1]:
# Pandas : is an open-source Python library used for data manipulation and analysis.
import pandas as pd

<h1 style="color:lightGreen; font-size:32px; font-weight:bold; background-color:#1C1C1C;padding:10px; border-radius:8px;">2.Loading the Dataset</h1>

##### <span style="color:orange"> importing data from an external file (such as CSV, Excel, JSON, or SQL) into a Pandas DataFrame so it can be analyzed and processed using Python. 

In [2]:
# Orders Table 
orders = pd.read_csv("olist_orders_dataset.csv")
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [3]:
# items Table
items = pd.read_csv('olist_order_items_dataset.csv')
items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [4]:
# payments Table
payments = pd.read_csv('olist_order_payments_dataset.csv')
payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

<h1 style="color:lightGreen; font-size:32px; font-weight:bold; background-color:#1C1C1C;padding:10px; border-radius:8px;">3.Combining Data</h1>

##### <span style="color:orange"> bringing together multiple datasets into one unified dataset

In [5]:
# Merges orders with items on order_id
orders_items = orders.merge(items , on = "order_id" , how = 'left')
orders_items.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0           2017-10-18 00:00:00            1.0   
1           2018-08-13 00:00:00            1.0   
2           2018-09-04 00:00:00            1.0   
3           2017-12-15 00:00:00            1.0   
4           2018-02-26 00:00:00            1.0   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
2  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   
3  d0b61bfb1de832b15ba9d266ca96e5b0  66922902710d126a0e7d26b0e3805106   
4  65266b2da20d04dbe00c5c2d3bb7859e  2c9e548be18521d1c43cde1c582c6de8   

   shipping_limit_date   price  freight_value  
0  2017-10-06 11:07:15   29.99           8.72  
1  2018-07-30 03:24:27  118.70          22.76  
2  2018-08-13 08:55:23  159.90          19.22  
3  2017-11-23 19:45:59   45.00          27.20  
4  2018-02-19 20:31:37   19.90           8.72

In [6]:
# Merges orders_items with Payments on order_id
df = orders_items.merge(payments, on="order_id", how="left")
df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  order_item_id  \
0           2017-10-18 00:00:00            1.0   
1           2017-10-18 00:00:00            1.0   
2           2017-10-18 00:00:00            1.0   
3           2018-08-13 00:00:00            1.0   
4           2018-09-04 00:00:00            1.0   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
2  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
3  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
4  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   

   shipping_limit_date   price  freight_value  payment_sequential  \
0  2017-10-06 11:07:15   29.99           8.72                 1.0   
1  2017-10-06 11:07:15   29.99           8.72                 3.0   
2  2017-10-06 11:07:15   29.99           8.72                 2.0   
3  2018-07-30 03:24:27  118.70          22.76                 1.0   
4  2018-08-13 08:55:23  159.90          19.22                 1.0   

  payment_type  payment_installments  payment_value  
0  credit_card                   1.0          18.12  
1      voucher                   1.0           2.00  
2      voucher                   1.0          18.59  
3       boleto                   1.0         141.46  
4  credit_card                   3.0         179.12

<h1 style="color:lightGreen; font-size:32px; font-weight:bold; background-color:#1C1C1C;padding:10px; border-radius:8px;">4.Exploring the Data</h1>

##### <span style="color:orange"> examining the dataset to understand its structure, quality, and key characteristics before performing analysis.

In [7]:
# Show the first 5 rows
df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  order_item_id  \
0           2017-10-18 00:00:00            1.0   
1           2017-10-18 00:00:00            1.0   
2           2017-10-18 00:00:00            1.0   
3           2018-08-13 00:00:00            1.0   
4           2018-09-04 00:00:00            1.0   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
2  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
3  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
4  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   

   shipping_limit_date   price  freight_value  payment_sequential  \
0  2017-10-06 11:07:15   29.99           8.72                 1.0   
1  2017-10-06 11:07:15   29.99           8.72                 3.0   
2  2017-10-06 11:07:15   29.99           8.72                 2.0   
3  2018-07-30 03:24:27  118.70          22.76                 1.0   
4  2018-08-13 08:55:23  159.90          19.22                 1.0   

  payment_type  payment_installments  payment_value  
0  credit_card                   1.0          18.12  
1      voucher                   1.0           2.00  
2      voucher                   1.0          18.59  
3       boleto                   1.0         141.46  
4  credit_card                   3.0         179.12

In [8]:
# Check data types and missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118434 entries, 0 to 118433
Data columns (total 18 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_id                       118434 non-null  object 
 1   customer_id                    118434 non-null  object 
 2   order_status                   118434 non-null  object 
 3   order_purchase_timestamp       118434 non-null  object 
 4   order_approved_at              118258 non-null  object 
 5   order_delivered_carrier_date   116360 non-null  object 
 6   order_delivered_customer_date  115037 non-null  object 
 7   order_estimated_delivery_date  118434 non-null  object 
 8   order_item_id                  117604 non-null  float64
 9   product_id                     117604 non-null  object 
 10  seller_id                      117604 non-null  object 
 11  shipping_limit_date            117604 non-null  object 
 12  price                         

In [9]:
# Description of the data in the DataFrame
df.describe()

order_item_id          price  freight_value  payment_sequential  \
count  117604.000000  117604.000000  117604.000000       118431.000000   
mean        1.195920     120.822849      20.045551            1.094207   
std         0.697721     184.477367      15.861351            0.728290   
min         1.000000       0.850000       0.000000            1.000000   
25%         1.000000      39.900000      13.080000            1.000000   
50%         1.000000      74.900000      16.290000            1.000000   
75%         1.000000     134.900000      21.190000            1.000000   
max        21.000000    6735.000000     409.680000           29.000000   

       payment_installments  payment_value  
count         118431.000000  118431.000000  
mean               2.937398     172.849395  
std                2.774164     268.259831  
min                0.000000       0.000000  
25%                1.000000      60.860000  
50%                2.000000     108.200000  
75%                4.000000     189.245000  
max               24.000000   13664.080000

In [10]:
# Checks for missing values 
df.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 176
order_delivered_carrier_date     2074
order_delivered_customer_date    3397
order_estimated_delivery_date       0
order_item_id                     830
product_id                        830
seller_id                         830
shipping_limit_date               830
price                             830
freight_value                     830
payment_sequential                  3
payment_type                        3
payment_installments                3
payment_value                       3
dtype: int64

In [11]:
# Checks for Duplicated values in Primary Key (order_id)
df['order_id'].duplicated().sum()

18993

<h1 style="color:#00274D; background-color:#FFD700; padding:10px; border-radius:8px;">
    📌 Featureset Exploration
</h1>

<ul style="color:#00FF7F; font-weight:bold; background-color:#2F2F2F; padding:20px; border-radius:5px;"> <b>
    <li>order_id : Nominal Data</li>
    <li>customer_id	 : Nominal Data</li>
    <li>order_status : Nominal Data (delivered , invoiced , shipped ,....) </li>
    <li style="color:red;">order_purchase_timestamp : Date Time [Note : Must be Date Time] </li> 
    <li style="color:red;">order_approved_at : Date Time [Note : Must be Date Time] </li>
    <li style="color:red;">order_delivered_carrier_date : Date Time [Note : Must be Date Time]</li>
    <li style="color:red;">order_delivered_customer_date : Date Time [Note : Must be Date Time]</li>
    <li style="color:red;">order_estimated_delivery_date : Date Time [Note : Must be Date Time]</li>
    <li style="color:red;">order_item_id : Integer [Note : Must be Integer]</li>
    <li>product_id : Nominal Data</li>
    <li>seller_id : Nominal Data</li>
    <li style="color:red;">shipping_limit_date	 : Date Time [Note : Must be Date Time]</li>
    <li>price : Continuous [29.99 , 45.00 , 38.25 , ....]</li>
    <li>freight_value : Continuous  [15.10, 30.53 , 8.7 ,  8.4 , .....] </li>
    <li style="color:red;">payment_sequential : Integer [Note : Must be Integer]</li>
    <li>payment_type : Nominal [credit_card , voucher , boleto , ...] </li>
    <li style="color:red;">payment_installments :  Integer [Note : Must be Integer]</li>
    <li>payment_value : Continuous [72.20 ,18.120 141.4625 , ....]</li>
</u>


<h1 style="color:lightGreen; font-size:32px; font-weight:bold; background-color:#1C1C1C;padding:10px; border-radius:8px;">4.Data Cleaning</h1>

##### <span style="color:orange"> is the process of identifying and correcting errors in a dataset to improve its quality and reliability. 

<ol style="color:#FF6347; font-weight:bold; background-color:#2F2F2F; padding:20px; border-radius:5px;"><b> 
    <li>Fix Duplicated Values(Order_id)</li>
     <br>
    <li>Handling Missing Values</li>
    <br>
    <li>Data Type Conversion</li>
    <br>
    <li>Creating New Useful Columns</li>
 </ol>

In [12]:
# Backing Up Data
df_clean = df.copy()
df_clean.head(4)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   

  order_estimated_delivery_date  order_item_id  \
0           2017-10-18 00:00:00            1.0   
1           2017-10-18 00:00:00            1.0   
2           2017-10-18 00:00:00            1.0   
3           2018-08-13 00:00:00            1.0   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
2  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
3  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   

   shipping_limit_date   price  freight_value  payment_sequential  \
0  2017-10-06 11:07:15   29.99           8.72                 1.0   
1  2017-10-06 11:07:15   29.99           8.72                 3.0   
2  2017-10-06 11:07:15   29.99           8.72                 2.0   
3  2018-07-30 03:24:27  118.70          22.76                 1.0   

  payment_type  payment_installments  payment_value  
0  credit_card                   1.0          18.12  
1      voucher                   1.0           2.00  
2      voucher                   1.0          18.59  
3       boleto                   1.0         141.46

<h3 style="color:#FF6347; font-size:22px; font-weight:bold; background-color:#2F2F2F; padding:5px; border-radius:5px;"">1-Fix Duplicated Values(Order_id)</h3>

##### <span style="color:orange"> Remove duplicate values in order_id (Primary Key)

In [13]:
df_clean['order_id'].duplicated().sum()

18993

In [14]:
# Removes duplicate rows based on order_id, keeping only the first occurrence
df_clean= df_clean.drop_duplicates(subset=['order_id'], keep='first')

In [15]:
df_clean['order_id'].duplicated().sum()

0

<h3 style="color:#FF6347; font-size:22px; font-weight:bold; background-color:#2F2F2F; padding:5px; border-radius:5px;"">2-Handling Missing Values</h3>

In [16]:
df_clean.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
order_item_id                     775
product_id                        775
seller_id                         775
shipping_limit_date               775
price                             775
freight_value                     775
payment_sequential                  1
payment_type                        1
payment_installments                1
payment_value                       1
dtype: int64

<h3 style="color:#FF6347; font-size:22px;"> 1. order_approved_at</h3>

In [17]:
# Chacks Null Values in order_approved_at
df_clean[df_clean["order_approved_at"].isna()].head()

order_id                       customer_id  \
1352  00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
2128  ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
2218  df8282afe61008dc26c6c31011474d02  aa797b187b5466bc6925aaaa4bb3bed1   
2415  8d4c637f1accf7a88a4555f02741e606  b1dd715db389a2077f43174e7a675d07   
2570  7a9d4c7f9b068337875b95465330f2fc  7f71ae48074c0cfec9195f88fcbfac55   

     order_status order_purchase_timestamp order_approved_at  \
1352     canceled      2018-08-28 15:26:39               NaN   
2128     canceled      2018-09-20 13:54:16               NaN   
2218     canceled      2017-03-04 12:14:30               NaN   
2415     canceled      2018-08-29 16:27:49               NaN   
2570     canceled      2017-05-01 16:12:39               NaN   

     order_delivered_carrier_date order_delivered_customer_date  \
1352                          NaN                           NaN   
2128                          NaN                           NaN   
2218                          NaN                           NaN   
2415                          NaN                           NaN   
2570                          NaN                           NaN   

     order_estimated_delivery_date  order_item_id product_id seller_id  \
1352           2018-09-12 00:00:00            NaN        NaN       NaN   
2128           2018-10-17 00:00:00            NaN        NaN       NaN   
2218           2017-04-10 00:00:00            NaN        NaN       NaN   
2415           2018-09-13 00:00:00            NaN        NaN       NaN   
2570           2017-05-30 00:00:00            NaN        NaN       NaN   

     shipping_limit_date  price  freight_value  payment_sequential  \
1352                 NaN    NaN            NaN                 1.0   
2128                 NaN    NaN            NaN                 1.0   
2218                 NaN    NaN            NaN                 1.0   
2415                 NaN    NaN            NaN                 1.0   
2570                 NaN    NaN            NaN                 1.0   

     payment_type  payment_installments  payment_value  
1352  not_defined                   1.0           0.00  
2128      voucher                   1.0         191.46  
2218       boleto                   1.0         139.96  
2415      voucher                   1.0          66.44  
2570  credit_card                   2.0          72.90

In [18]:
# Fills missing values in order_approved_at with values from order_purchase_timestamp
df_clean['order_approved_at'] = df_clean['order_approved_at'].fillna(df_clean['order_purchase_timestamp'])

<h3 style="color:#FF6347; font-size:22px;"> 2. order_delivered_carrier_date </h3>

In [19]:
df_clean[df_clean["order_delivered_carrier_date"].isna()].head()

order_id                       customer_id  \
8    136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
125  0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
155  15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
319  8e24261a7e58791d10cb1bf9da94df5c  64a254d30eed42cd0e6c36dddb88adf0   
385  d3c8851a6651eeff2f73b0e011ac45d0  957f8e082185574de25992dc659ebbc0   

    order_status order_purchase_timestamp    order_approved_at  \
8       invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
125     invoiced      2018-08-03 17:44:42  2018-08-07 06:15:14   
155   processing      2017-09-03 14:22:03  2017-09-03 14:30:09   
319  unavailable      2017-11-16 15:09:28  2017-11-16 15:26:57   
385   processing      2016-10-05 22:44:13  2016-10-06 15:51:05   

    order_delivered_carrier_date order_delivered_customer_date  \
8                            NaN                           NaN   
125                          NaN                           NaN   
155                          NaN                           NaN   
319                          NaN                           NaN   
385                          NaN                           NaN   

    order_estimated_delivery_date  order_item_id  \
8             2017-05-09 00:00:00            1.0   
125           2018-08-21 00:00:00            1.0   
155           2017-10-03 00:00:00            1.0   
319           2017-12-05 00:00:00            NaN   
385           2016-12-09 00:00:00            1.0   

                           product_id                         seller_id  \
8    a1804276d9941ac0733cfd409f5206eb  dc8798cbf453b7e0f98745e396cc5616   
125  1522589c64efd46731d3522568e5bc83  28405831a29823802aa22c084cfd0649   
155  61d52f4882421048afd530db53d6f230  fa74b2f3287d296e9fbd2cc80f2d1cf1   
319                               NaN                               NaN   
385  ca17b18b38db6e0d2c33d362322f2a06  4be2e7f96b4fd749d52dff41f80e39dd   

     shipping_limit_date  price  freight_value  payment_sequential  \
8    2017-04-19 13:25:17   49.9          16.05                 1.0   
125  2018-08-13 06:15:14   35.0          15.35                 1.0   
155  2017-09-20 14:30:09  125.9          12.38                 1.0   
319                  NaN    NaN            NaN                 1.0   
385  2016-10-21 16:24:45  267.0          32.84                 1.0   

    payment_type  payment_installments  payment_value  
8    credit_card                   1.0          65.95  
125       boleto                   1.0          50.35  
155  credit_card                   2.0         138.28  
319  credit_card                   5.0          84.00  
385  credit_card                   1.0         299.84

In [20]:
# Fills missing values in order_delivered_carrier_date with values from order_purchase_timestamp
df_clean['order_delivered_carrier_date'] = df_clean['order_delivered_carrier_date'].fillna(df_clean['order_purchase_timestamp'])

<h3 style="color:#FF6347; font-size:22px;"> 3. order_delivered_customer_date </h3>

In [21]:
df_clean[df_clean["order_delivered_customer_date"].isna()].head()

order_id                       customer_id  \
8    136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
51   ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
125  0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
155  15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
192  6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   

    order_status order_purchase_timestamp    order_approved_at  \
8       invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
51       shipped      2018-06-04 16:44:48  2018-06-05 04:31:18   
125     invoiced      2018-08-03 17:44:42  2018-08-07 06:15:14   
155   processing      2017-09-03 14:22:03  2017-09-03 14:30:09   
192      shipped      2018-01-10 11:33:07  2018-01-11 02:32:30   

    order_delivered_carrier_date order_delivered_customer_date  \
8            2017-04-11 12:22:08                           NaN   
51           2018-06-05 14:32:00                           NaN   
125          2018-08-03 17:44:42                           NaN   
155          2017-09-03 14:22:03                           NaN   
192          2018-01-11 19:39:23                           NaN   

    order_estimated_delivery_date  order_item_id  \
8             2017-05-09 00:00:00            1.0   
51            2018-06-28 00:00:00            1.0   
125           2018-08-21 00:00:00            1.0   
155           2017-10-03 00:00:00            1.0   
192           2018-02-07 00:00:00            1.0   

                           product_id                         seller_id  \
8    a1804276d9941ac0733cfd409f5206eb  dc8798cbf453b7e0f98745e396cc5616   
51   c50ca07e9e4db9ea5011f06802c0aea0  e9779976487b77c6d4ac45f75ec7afe9   
125  1522589c64efd46731d3522568e5bc83  28405831a29823802aa22c084cfd0649   
155  61d52f4882421048afd530db53d6f230  fa74b2f3287d296e9fbd2cc80f2d1cf1   
192  ee0c1cf2fbeae95205b4aa506f1469f0  cc419e0650a3c5ba77189a1882b7556a   

     shipping_limit_date   price  freight_value  payment_sequential  \
8    2017-04-19 13:25:17   49.90          16.05                 1.0   
51   2018-06-13 04:30:33   14.49           7.87                 1.0   
125  2018-08-13 06:15:14   35.00          15.35                 1.0   
155  2017-09-20 14:30:09  125.90          12.38                 1.0   
192  2018-01-18 02:32:30   53.99          15.13                 1.0   

    payment_type  payment_installments  payment_value  
8    credit_card                   1.0          65.95  
51        boleto                   1.0          22.36  
125       boleto                   1.0          50.35  
155  credit_card                   2.0         138.28  
192       boleto                   1.0          69.12

In [22]:
# Fills missing values in order_delivered_customer_date with values from order_estimated_delivery_date
df_clean['order_delivered_customer_date'] = df_clean['order_delivered_customer_date'].fillna(df_clean['order_estimated_delivery_date'])

<h3 style="color:#FF6347; font-size:22px;"> 4. order_item_id</h3>

In [23]:
df_clean[df_clean['order_item_id'].isnull()].head()

order_id                       customer_id  \
319   8e24261a7e58791d10cb1bf9da94df5c  64a254d30eed42cd0e6c36dddb88adf0   
700   c272bcd21c287498b4883c7512019702  9582c5bbecc65eb568e2c1d839b5cba1   
821   37553832a3a89c9b2db59701c357ca67  7607cd563696c27ede287e515812d528   
880   d57e15fb07fd180f06ab3926b39edcd2  470b93b3f1cde85550fc74cd3a476c78   
1352  00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   

     order_status order_purchase_timestamp    order_approved_at  \
319   unavailable      2017-11-16 15:09:28  2017-11-16 15:26:57   
700   unavailable      2018-01-31 11:31:37  2018-01-31 14:23:50   
821   unavailable      2017-08-14 17:38:02  2017-08-17 00:15:18   
880   unavailable      2018-01-08 19:39:03  2018-01-09 07:26:08   
1352     canceled      2018-08-28 15:26:39  2018-08-28 15:26:39   

     order_delivered_carrier_date order_delivered_customer_date  \
319           2017-11-16 15:09:28           2017-12-05 00:00:00   
700           2018-01-31 11:31:37           2018-02-16 00:00:00   
821           2017-08-14 17:38:02           2017-09-05 00:00:00   
880           2018-01-08 19:39:03           2018-02-06 00:00:00   
1352          2018-08-28 15:26:39           2018-09-12 00:00:00   

     order_estimated_delivery_date  order_item_id product_id seller_id  \
319            2017-12-05 00:00:00            NaN        NaN       NaN   
700            2018-02-16 00:00:00            NaN        NaN       NaN   
821            2017-09-05 00:00:00            NaN        NaN       NaN   
880            2018-02-06 00:00:00            NaN        NaN       NaN   
1352           2018-09-12 00:00:00            NaN        NaN       NaN   

     shipping_limit_date  price  freight_value  payment_sequential  \
319                  NaN    NaN            NaN                 1.0   
700                  NaN    NaN            NaN                 1.0   
821                  NaN    NaN            NaN                 1.0   
880                  NaN    NaN            NaN                 1.0   
1352                 NaN    NaN            NaN                 1.0   

     payment_type  payment_installments  payment_value  
319   credit_card                   5.0          84.00  
700   credit_card                   1.0          97.68  
821        boleto                   1.0         132.46  
880        boleto                   1.0         134.38  
1352  not_defined                   1.0           0.00

In [24]:
# Fills missing values in order_item_id with 0
df_clean['order_item_id'] = df_clean['order_item_id'].fillna(0)

<h3 style="color:#FF6347; font-size:22px;"> 5. product_id , seller_id </h3>

In [25]:
# Fills missing values in product_id and seller_id with 'unknown'
df_clean[['product_id' ,'seller_id']] = df_clean[['seller_id','product_id']].fillna('unknown')

<h3 style="color:#FF6347; font-size:22px;"> 5. shipping_limit_date </h3>

In [26]:
# Converts shipping_limit_date to datetime format
df_clean['shipping_limit_date'] = pd.to_datetime(df_clean['shipping_limit_date'])

In [27]:
# Converts order_estimated_delivery_date to datetime format
df_clean['order_estimated_delivery_date'] = pd.to_datetime(df_clean['order_estimated_delivery_date'])

In [28]:
# Fills missing values in shipping_limit_date with order_estimated_delivery_date - 5 days
df_clean['shipping_limit_date'] = df_clean['shipping_limit_date'].fillna(df_clean['order_estimated_delivery_date']  - pd.Timedelta(days=5))

In [29]:
df_clean.isnull().sum()

order_id                           0
customer_id                        0
order_status                       0
order_purchase_timestamp           0
order_approved_at                  0
order_delivered_carrier_date       0
order_delivered_customer_date      0
order_estimated_delivery_date      0
order_item_id                      0
product_id                         0
seller_id                          0
shipping_limit_date                0
price                            775
freight_value                    775
payment_sequential                 1
payment_type                       1
payment_installments               1
payment_value                      1
dtype: int64

<h3 style="color:#FF6347; font-size:22px;"> 6. Price , freight_value</h3>

In [30]:
# Fills missing values in price and freight_value with 0
df_clean[['price' , 'freight_value']] = df_clean[['price' , 'freight_value']].fillna(0)

In [31]:
df_clean.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
payment_sequential               1
payment_type                     1
payment_installments             1
payment_value                    1
dtype: int64

<h3 style="color:#FF6347; font-size:22px;"> 7. Price , freight_value</h3>

In [32]:
# Drops rows where any of payment_sequential, payment_type, payment_installments, or payment_value has missing values
df_clean = df_clean.dropna(subset=['payment_sequential', 'payment_type', 'payment_installments', 'payment_value'])

In [33]:
df_clean.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
payment_sequential               0
payment_type                     0
payment_installments             0
payment_value                    0
dtype: int64

<h3 style="color:#FF6347; font-size:22px; font-weight:bold; background-color:#2F2F2F; padding:5px; border-radius:5px;"">3-Data Type Conversion</h3>

##### <span style="color:orange"> Data Type Conversion refers to the process of changing the data type of a variable or column in a dataset to another type. 

<h3 style="color:#FF6347; font-size:22px;"> 1. order_purchase_timestamp (From object  to Date Time)</h3>

In [34]:
# returns the data type
df_clean['order_purchase_timestamp'].dtype

dtype('O')

In [35]:
# To convert a Time column from object (Object) to datetime format, use pd.to_datetime().
df_clean['order_purchase_timestamp'] = pd.to_datetime(df_clean['order_purchase_timestamp'])

In [36]:
df_clean['order_purchase_timestamp'].dtype

dtype('<M8[ns]')

<h3 style="color:#FF6347; font-size:22px;"> 2. order_approved_at (From object  to Date Time)</h3>

In [37]:
df_clean['order_approved_at'].dtype

dtype('O')

In [38]:
df_clean['order_approved_at'] = pd.to_datetime(df_clean['order_approved_at'])

In [39]:
df_clean['order_approved_at'].dtype

dtype('<M8[ns]')

<h3 style="color:#FF6347; font-size:22px;"> 3. order_delivered_carrier_date (From object  to Date Time)</h3>

In [40]:
df_clean['order_delivered_carrier_date'].dtype

dtype('O')

In [41]:
df_clean['order_delivered_carrier_date'] = pd.to_datetime(df_clean['order_delivered_carrier_date'])

In [42]:
df_clean['order_delivered_carrier_date'].dtype

dtype('<M8[ns]')

<h3 style="color:#FF6347; font-size:22px;"> 4. order_delivered_customer_date (From object  to Date Time)</h3>

In [43]:
df_clean['order_delivered_customer_date'].dtype

dtype('O')

In [44]:
df_clean['order_delivered_customer_date'] = pd.to_datetime(df_clean['order_delivered_customer_date'])

In [45]:
df_clean['order_delivered_customer_date'].dtype

dtype('<M8[ns]')

<h3 style="color:#FF6347; font-size:22px;"> 5. order_item_id , payment_sequential , payment_installments (From Float to int)</h3>

In [46]:
df_clean[['order_item_id','payment_sequential','payment_installments']] = df_clean[['order_item_id','payment_sequential',
                    'payment_installments']].astype(int)

In [47]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99440 entries, 0 to 118433
Data columns (total 18 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99440 non-null  object        
 1   customer_id                    99440 non-null  object        
 2   order_status                   99440 non-null  object        
 3   order_purchase_timestamp       99440 non-null  datetime64[ns]
 4   order_approved_at              99440 non-null  datetime64[ns]
 5   order_delivered_carrier_date   99440 non-null  datetime64[ns]
 6   order_delivered_customer_date  99440 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99440 non-null  datetime64[ns]
 8   order_item_id                  99440 non-null  int32         
 9   product_id                     99440 non-null  object        
 10  seller_id                      99440 non-null  object        
 11  shipping_limit_date

In [48]:
df_clean.describe()

order_purchase_timestamp              order_approved_at  \
count                          99440                          99440   
mean   2017-12-31 08:50:02.752393472  2017-12-31 19:14:11.517125632   
min              2016-09-04 21:15:19            2016-10-04 09:43:32   
25%       2017-09-12 14:50:17.500000            2017-09-12 21:55:19   
50%              2018-01-18 23:05:58            2018-01-19 11:54:31   
75%              2018-05-04 15:42:52  2018-05-04 23:50:21.249999872   
max              2018-10-17 17:30:18            2018-10-17 17:30:18   
std                              NaN                            NaN   

        order_delivered_carrier_date  order_delivered_customer_date  \
count                          99440                          99440   
mean   2018-01-03 13:02:51.128409088  2018-01-13 06:56:54.560086272   
min              2016-09-05 00:15:34            2016-09-30 00:00:00   
25%              2017-09-14 20:18:42  2017-09-22 21:28:59.249999872   
50%              2018-01-23 16:49:35            2018-02-01 19:34:52   
75%              2018-05-08 08:47:00  2018-05-15 17:51:04.249999872   
max              2018-10-17 17:30:18            2018-11-12 00:00:00   
std                              NaN                            NaN   

       order_estimated_delivery_date  order_item_id  \
count                          99440   99440.000000   
mean   2018-01-24 03:15:32.292839680       0.992206   
min              2016-09-30 00:00:00       0.000000   
25%              2017-10-03 00:00:00       1.000000   
50%              2018-02-15 00:00:00       1.000000   
75%              2018-05-25 00:00:00       1.000000   
max              2018-11-12 00:00:00       1.000000   
std                              NaN       0.087937   

              shipping_limit_date         price  freight_value  \
count                       99440  99440.000000   99440.000000   
mean   2018-01-07 01:37:47.473280    124.983420      20.044655   
min           2016-09-19 00:15:34      0.000000       0.000000   
25%    2017-09-19 17:45:04.500000     39.990000      13.127500   
50%           2018-01-25 17:50:16     78.900000      16.320000   
75%    2018-05-10 13:10:04.500000    139.890000      21.180000   
max           2020-04-09 22:35:08   6735.000000     409.680000   
std                           NaN    190.950239      15.947000   

       payment_sequential  payment_installments  payment_value  
count        99440.000000          99440.000000   99440.000000  
mean             1.022878              2.915617     158.334781  
min              1.000000              0.000000       0.000000  
25%              1.000000              1.000000      60.050000  
50%              1.000000              2.000000     103.350000  
75%              1.000000              4.000000     175.130000  
max             27.000000             24.000000   13664.080000  
std              0.250282              2.709861     219.518488

In [49]:
df_clean.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
payment_sequential               0
payment_type                     0
payment_installments             0
payment_value                    0
dtype: int64

<h3 style="color:#FF6347; font-size:22px; font-weight:bold; background-color:#2F2F2F; padding:5px; border-radius:5px;"">4-Creating New Useful Columns</h3>

<ol style="color:#FF6347; font-weight:bold; background-color:#2F2F2F; padding:20px; border-radius:5px;"><b> 
    <li>total_revenue = price + freight_value</li>
     <br>
    <li>net_profit = price - freight_value</li>
    <br>
    <li>expected_revenue = order_status == processing , shipped , approved , created</li>
    <br>
    <li>canceled_orders = (order_status == 'canceled)</li>
    <br>
    <li>late_deliveries = order_delivered_customer_date > order_estimated_delivery_date</li>
 </ol>

<h3 style="color:#FF6347; font-size:22px;"> 1. total_revenue</h3>

In [50]:
# total_revenue = price + freight_value
df_clean['total_revenue'] = df_clean['price'] + df_clean['freight_value']

In [51]:
# Moves the total_revenue column to position 14 in df_clean
df_clean.insert(14,'total_revenue', df_clean.pop('total_revenue'))

In [52]:
df_clean.head(4)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
5  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
5    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   
5          2017-11-22 13:39:59           2017-12-02 00:28:42   

  order_estimated_delivery_date  order_item_id  \
0                    2017-10-18              1   
3                    2018-08-13              1   
4                    2018-09-04              1   
5                    2017-12-15              1   

                         product_id                         seller_id  \
0  3504c0cb71d7fa48d967e0e4c94d59d9  87285b34884572647811a353c7ac498a   
3  289cdb325fb7e7f891c38608bf9e0962  595fac2a385ac33a80bd5114aec74eb8   
4  4869f7a5dfa277a7dca6462dcf3b52b2  aa4383b373c6aca5d8797843e5594415   
5  66922902710d126a0e7d26b0e3805106  d0b61bfb1de832b15ba9d266ca96e5b0   

  shipping_limit_date   price  freight_value  total_revenue  \
0 2017-10-06 11:07:15   29.99           8.72          38.71   
3 2018-07-30 03:24:27  118.70          22.76         141.46   
4 2018-08-13 08:55:23  159.90          19.22         179.12   
5 2017-11-23 19:45:59   45.00          27.20          72.20   

   payment_sequential payment_type  payment_installments  payment_value  
0                   1  credit_card                     1          18.12  
3                   1       boleto                     1         141.46  
4                   1  credit_card                     3         179.12  
5                   1  credit_card                     1          72.20

<h3 style="color:#FF6347; font-size:22px;"> 2. net_profit</h3>

In [53]:
#net_profit = price - freight_value
df_clean['net_profit'] = df_clean['price'] - df_clean['freight_value']

In [54]:
# Moves the net_profit column to position 14 in df_clean
df_clean.insert(14,'net_profit', df_clean.pop('net_profit'))

In [55]:
df_clean.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
5  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
6  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
5    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
6    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   
5          2017-11-22 13:39:59           2017-12-02 00:28:42   
6          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0                    2017-10-18              1   
3                    2018-08-13              1   
4                    2018-09-04              1   
5                    2017-12-15              1   
6                    2018-02-26              1   

                         product_id                         seller_id  \
0  3504c0cb71d7fa48d967e0e4c94d59d9  87285b34884572647811a353c7ac498a   
3  289cdb325fb7e7f891c38608bf9e0962  595fac2a385ac33a80bd5114aec74eb8   
4  4869f7a5dfa277a7dca6462dcf3b52b2  aa4383b373c6aca5d8797843e5594415   
5  66922902710d126a0e7d26b0e3805106  d0b61bfb1de832b15ba9d266ca96e5b0   
6  2c9e548be18521d1c43cde1c582c6de8  65266b2da20d04dbe00c5c2d3bb7859e   

  shipping_limit_date   price  freight_value  net_profit  total_revenue  \
0 2017-10-06 11:07:15   29.99           8.72       21.27          38.71   
3 2018-07-30 03:24:27  118.70          22.76       95.94         141.46   
4 2018-08-13 08:55:23  159.90          19.22      140.68         179.12   
5 2017-11-23 19:45:59   45.00          27.20       17.80          72.20   
6 2018-02-19 20:31:37   19.90           8.72       11.18          28.62   

   payment_sequential payment_type  payment_installments  payment_value  
0                   1  credit_card                     1          18.12  
3                   1       boleto                     1         141.46  
4                   1  credit_card                     3         179.12  
5                   1  credit_card                     1          72.20  
6                   1  credit_card                     1          28.62

<h3 style="color:#FF6347; font-size:22px;"> 3. expected_revenue</h3>

In [56]:
# Calculates expected_revenue as price + freight_value only for orders with status approved, processing, or shipped; otherwise, sets it to 0
df_clean['expected_revenue'] = df_clean.apply(
    lambda x: x['price'] + x['freight_value'] if x['order_status'] in ['approved', 'processing', 'shipped'] else 0, axis=1
)

In [57]:
df_clean.groupby('order_status')['expected_revenue'].sum()

order_status
approved          210.27
canceled            0.00
created             0.00
delivered           0.00
invoiced            0.00
processing      60648.75
shipped        169347.24
unavailable         0.00
Name: expected_revenue, dtype: float64

<h3 style="color:#FF6347; font-size:22px;"> 4. canceled_orders </h3>

In [58]:
# Creates a new column canceled_orders (order_status == 'canceled')
df_clean['canceled_orders'] = df_clean.apply(
    lambda x: 1 if x['order_status'] == 'canceled' else 0, axis=1
)

In [59]:
total_late_deliveries = df_clean['canceled_orders'].sum()
print("Total Canceled Orders:", total_late_deliveries)

Total Canceled Orders: 625


<h3 style="color:#FF6347; font-size:22px;"> 5. late_deliveries</h3>

In [60]:
# Marks late deliveries 1 if delivered late, 0 otherwise
df_clean['late_deliveries'] = df_clean.apply(
    lambda x: 1 if x['order_delivered_customer_date'] > x['order_estimated_delivery_date'] else 0, axis=1)


In [61]:
total_late_deliveries = df_clean['late_deliveries'].sum()
print("Total Late Deliveries:", total_late_deliveries)

Total Late Deliveries: 7826


In [62]:
df_clean.insert(16,'expected_revenue' , df_clean.pop('expected_revenue'))

In [63]:
df_clean.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
5  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
6  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
5    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
6    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   
5          2017-11-22 13:39:59           2017-12-02 00:28:42   
6          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0                    2017-10-18              1   
3                    2018-08-13              1   
4                    2018-09-04              1   
5                    2017-12-15              1   
6                    2018-02-26              1   

                         product_id  ... freight_value net_profit  \
0  3504c0cb71d7fa48d967e0e4c94d59d9  ...          8.72      21.27   
3  289cdb325fb7e7f891c38608bf9e0962  ...         22.76      95.94   
4  4869f7a5dfa277a7dca6462dcf3b52b2  ...         19.22     140.68   
5  66922902710d126a0e7d26b0e3805106  ...         27.20      17.80   
6  2c9e548be18521d1c43cde1c582c6de8  ...          8.72      11.18   

   total_revenue  expected_revenue  payment_sequential  payment_type  \
0          38.71               0.0                   1   credit_card   
3         141.46               0.0                   1        boleto   
4         179.12               0.0                   1   credit_card   
5          72.20               0.0                   1   credit_card   
6          28.62               0.0                   1   credit_card   

   payment_installments  payment_value canceled_orders  late_deliveries  
0                     1          18.12               0                0  
3                     1         141.46               0                0  
4                     3         179.12               0                0  
5                     1          72.20               0                0  
6                     1          28.62               0                0  

[5 rows x 23 columns]

In [64]:
# Save File 'df_clean'
# df_clean.to_csv("Brazilian E-Commerce Clean.csv" ,index=False )